src/poll.rs

```diff
@@ -52,14 +90,21 @@ const MAX_REFCOUNT: usize = (isize::MAX) as usize;
 /// poll.poll(&mut events, None).unwrap();
 /// ```
 pub struct Poll {
-    // This type is `Send`, but not `Sync`, so ensure it's exposed as such.
-    _marker: marker::PhantomData<Cell<()>>,
-
     // Platform specific IO selector
     selector: sys::Selector,
 
     // Custom readiness queue
     readiness_queue: ReadinessQueue,
+
+    // Use an atomic to first check if a full lock will be required. This is a
+    // fast-path check for single threaded cases avoiding the extra syscall
+    lock_state: AtomicUsize,
```

Comment:
I ran the benchmark with this lock_state “optimisation” naively removed and it was actually a tiny bit faster than when this fast-path is present on my system.